design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+
+## What is not in scope
+
+- Address unavailability caused by cold starts of KSQL server i.e when a new server is added to a 
+KSQL cluster, it must first rebuild its state off the changelog topics and that process still could 
+take a long time. 
+- Try to improve consistency guarantees provided by KSQL i.e reduce the amount of time it takes to 
+rebalance or standby replication to catch up. 
+
+## Value/Return
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+No change to public APIs
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
+Healthchecking, Lag reporting and Lag-aware routing.  
+
+### Failure detection/Healthcheck
+
+Failure detection/ healthcheck is implemented via a periodic hearbeat. KSQL servers either broadcast 
+their heartbeat (N^2 interconnections with N KSQL servers) or we implement a gossip protocol. In 
+the initial implementation, we will use the REST API to send heartbeats leveraging the N^2 mesh that 
+already exists between KSQL servers. Hence, we will implement a new REST endpoint, `/heartbeat` that 
+will register the heartbeats a server receives from other servers. 
+
+We want to guarantee 99% uptime. This means that in a 5 minute window, it must take no longer than 
+2 seconds to determine the active server is down and to forward the request to one of the standbys. 
+The heartbeats must be light-weight so that we can send multiple heartbeats per second which will 
+provide us with more datapoints required to implement a well-informed policy for determining when a 
+server is up and when down.
+
+Configuration parameters:
+1) Heartbeat INTERVAL (e.g send heartbeat every 200 ms)
+2) Heartbeat WINDOW (e.g every 2 seconds, count the missed/received heartbeats and determine if server is down/up)
+3) MISSED_THRESHOLD: How many heartbeats in a row constitute a node as down (e.g. 3 missed heartbeats = server is down)
+4) RECEIVED_THRESHOLD: How many heartbeats in a row constitute a node as up (e.g. 2 received heartbeats = server is up)
+
+We will provide sane defaults out-of-box, to achieve a 99% uptime.
+
+Pseudocode for REST endpoint for `/heartbeat`:
+```java
+Map<KsqlServer, List<Long>> receivedHeartbeats;
+
+@POST
+public Response receiveHeartbeat(Request heartbeat) 
+  return Response.ok(processRequest()).build();
+}
+
+private HeartbeatResponse processRequest() {
+  KsqlServer server = request.getServer();
+  receivedHeartbeats.get(server).add(System.currentTimeMillis());
+}
+```
+
+
+### Lag reporting
```

Comment:
Rather than just lag, how about reporting offsets (either current offset + lag, or current offset + end offset)? That way, if we wanted to we could incorporate offset into the request handling (e.g. route me to a secondary that has x lag as long as it's not earlier than some offset. This way clients can ensure their view of state doesn't go backwards in time.

Reply:
I am not sure if we have a direct use-case where the queries express temporal or offset based constraints. So we chose to keep this simpler on the Streams side for now. Just route this information to the most caught up replica. 

We can explore these as the actual use-case comes up?

Reply:
Reporting lag by itself may be dangerous as the lag is relative to the head offset, which isn't available to the node using the lag information.

For example, consider:

1. head offset is `100`
1. `A` reports its lag to `C` as `10`, as its only at offset `90`
1. head offset moves to `300`
1. `B` reports its lag to `C` as `20`, as its only at offset `290`.

Now if `C` tries to compute the most up-to-date node to use, based on the lag, it will choose `A`, even though its not as up-to-date as `B`.

I would think what we want to communicate is the offset information:

1. head offset is `100`
1. `A` reports its offset to `C` as `90`.
1. head offset moves to `300`
1. `B` reports its offset to `C` as `290`

Now `C` can correctly determine that `B` is more up-to-date than `A`.

Reply:
Good point. We also discussed this before inside Streams team. For C to determine who's upto date, it needs to fetch the head offset by itself for all partitions as some cadence.. So depending on how that gets scheduled and how fresh of a `head` value `C` uses (100, 300), you can also get positive or negative lag for B. 

We could send both the offset and head `A sends (90,100)` and `B sends (290, 300)`, to avoid this sort of scenario. But still since each node only refreshes this lag information every few seconds, `C` can still decide based on stale data and incorrectly route if say in the meantime `A` has reached the offset `299` on head `300`. i.e it caught up in the lag propagation time period. 

The real issue here, I think is the out-of-date head position. We have some ideas around solving this in Kafka/Streams, and then piggy back a much more real-time lag along with.
For now, our goal here is to provide some sort of basic fencing to avoid querying state that is lagging by a lot, for which the current approach seems fine IMO. 

Sending both `(offset, head)` seems more future proof, this needs a change in the KIP/discussion. We can decide based on how that goes.

Reply:
Tracked by issue #4227